functions/go/set-image/README.md

```diff
@@ -0,0 +1,118 @@
+# set-image
+
+## Overview
+
+<!--mdtogo:Short-->
+
+The `set-image` function sets the image for all instances of a provided image
+`name` with the provided image `newName:newTag` or `newName@digest`. A common
+deployment pattern is to update images to the latest tags or, ideally, digests.
+Sometimes people want to change the image repo, also, such as for dev builds.
+
+<!--mdtogo-->
+
+You can learn more about images [here][image].
+
+<!--mdtogo:Long-->
+
+## Usage
+
+This function can be used with any KRM function orchestrators (e.g. kpt).
+
+For all resources which specify an image of a given `name`, the `set-image`
+function sets that image field to the specified `newName:newTag` or
+`newName@digest`.
+
+By default the function updates the standard [fields][commonimage] used to
+specify an image.
+
+This function can be used both declaratively and imperatively.
+
+### FunctionConfig
+
+There are 2 kinds of `functionConfig` supported by this function:
+
+- `ConfigMap`
+- A custom resource of kind `SetImage`
+
+To use a `ConfigMap` as the `functionConfig`, the desired image spec must be
+specified using the following fields:
+- `data.name`
+- `data.newName`
```

Comment:
If I don't want to update the image name, do I still need to specify this field?

Reply:
In the current implementation, yes you still need to supply a `newName`. However to leave the name unchanged you can specify the same value for both `name` and `newName`

Reply:
In the current design, `newName` seems to be a required field, but IMO it doesn't have to be.
I guess we can make it optional: if it's specified, we update the image. If not, we keep the image name unchanged.

Reply:
This is a good point, and has me rethinking the input validation that I've put in place here.. I put the input validation in hopes of providing a more intuitive UX, however upon reconsideration it seems the default behavior of the kyaml filter may be desirable here. 

For reference, here is the default behavior of the kyaml filter:
`name`: the original name used to identify which images to replace. If omitted the operation becomes a noop
`newName`: the new name to set for images matching `name`. If omitted the name will be unchanged
`newTag`: the new tag to set for images matching `name`. If omitted the tag will be unchanged
`digest`: the new digest to set for images matching `name`. if omitted the digest will be unchanged. If provided, the `digest` will be used rather than `newTag`

I think we could largely just rely on the default behavior here, but I'd like to suggest a few exceptions which we can validate ourselves:
- Make `name` a required field, since the function becomes a noop without a provided `name`
- Allow `newName`, `newTag`, and `digest` to be optional. But return a validation error if none are provided since the function will become a noop
- Return a validation error if both a `newTag` and a `digest` are provided. If a `newTag` and `digest` are provided then the kyaml filter will silently ignore the `newTag` and use `digest`, which seems like a potentially confusing function UX.

WDYT?

Reply:
The default behavior of the kyaml filter + the validation logic you suggested SGTM

Reply:
Done